server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,131 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.get;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TransportGetFromTranslogAction extends HandledTransportAction<GetRequest, TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, GetRequest::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetRequest request, ActionListener<Response> listener) {
+        final ShardId shardId = request.getInternalShardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert request.realtime();
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
+        var get = new Engine.Get(request.realtime(), request.realtime(), request.id()).version(request.version())
+            .versionType(request.versionType());
+        assert Objects.equals(get.uid().field(), IdFieldMapper.NAME) : get.uid().field();
+        ActionListener.completeWith(listener, () -> {
+            if (engine.isInVersionMap(get.uid().bytes())) {
```

Comment:
I was wondering if maybe this should just be implemented as an internal detail of `InternalEngine#get` - like a parameter on `GetResult` that is get from translog only? And it sidesteps the searcher fallback in that method?

Would make it all operate from the single call to the version map. And avoids exposing more methods on the `Engine` which is suboptimal.

The return method if it is not found in translog could be `GetResult.NOT_EXISTS` or it could be expanded a bit to clarify that only the translog was checked.

@henningandersen thoughts?

Reply:
Btw - this comment in `InternalEngine` is out of date right? We definitely read non-updates from translog these days I think.

```
if (get.isReadFromTranslog()) {
                        // this is only used for updates - API _GET calls will always read form a reader for consistency
                        // the update call doesn't need the consistency since it's source only + _parent but parent can go away in 7.0
```

Reply:
Yes. I was thinking the same after Henning's comment. Returning early from `get` is a way to also address what Henning mentioned above since we'll do everything in one go. I'm ok with doing it here or a follow up.

Reply:
And yes, that comment is outdated. `get.isReadFromTranslog` gets the value of `get.realtime` except for one place I think (`TermVectorsService`).

Reply:
I like Tim's proposal. Whether it is done here or in a follow-up is fine by me.

Yes, the comment is out of data, I've stumbled upon that a few times myself too. Updating it in this PR makes sense to me.

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/93976/commits/3c6f60ac1a439e94c517cd1ef91038d1227b0b47